Create ocflib Tools for Announcements

vicshi06 opened the pull request.

vicshi06 commented:
@ben9583 Can you take a look at the first draft? I am able to use the GitHub API to get the content and the metadata of one specific announcement based on the ID. I will work on the third feature later. Here is the output for the functions. 

## Content 
```bash
---
title: "testing"
date: 2002-01-00
author: Victor
tags: [test, test2, test3]
summary: this is a test announcement 
---

# Test Announcement

## Summary

- this is a test announcement.

## Detail

- detail of the test annoucement.

## Contacts

- Name: Victor
- Email: contact@berkeley.edu
- Phone: (123)456-7890

```

## Metadata
```bash
title: "testing"
date: 2002-01-00
author: Victor
tags: [test, test2, test3]
summary: this is a test announcement 
```

vicshi06 commented:
@ben9583 Hi Ben. Can you take a look and make sure the function behaviors are good? I will implement cache and make the code more efficient later.

ben9583 commented on ocflib/lab/announcements.py:
```diff
@@ -0,0 +1,71 @@
+    return [get_announcement(get_id(post)) for post in posts[-1 : -num - 1 : -1]]
+
+
+# print(get_all_announcements())
```
Let's get rid of these debug statements for the pr

ben9583 commented on ocflib/lab/announcements.py:
```diff
@@ -0,0 +1,71 @@
+
+def get_id(post: str) -> str:
+    """Get announcement id"""
+    return post["name"][:-3]
```
Can you enumerate more why this works? Just so that way other people know what's going on here.

vicshi06 replied:
Since the name of the announcements is `YYYY-MM-DD.md`, this is to grab everything except for the last 3 characters. I can put more comments in the code as well.

ben9583 replied:
Yea I remember, but I think comments in code would be good.

ben9583 commented:
As for why the tests are failing, one of them is a regular housekeeping task (don't worry about it), and the other two are... from my code from 2 years ago. I'll take a look at why they're failing now.

ben9583 commented:
Ok I'm blaming the issues with mine on daylight savings happening that day. There only seems to be one commit in which it fails and that doesn't include the most recent one

vicshi06 commented:
Ok is there anything I need to do to fix the checks?

ben9583 commented:
Will take a more thorough look at this tomorrow, but it seems the [tests on Jenkins](https://jenkins.ocf.berkeley.edu/blue/organizations/jenkins/ocf%2Focflib/detail/PR-257/7/pipeline/30/#step-33-log-1) are not working properly since the recent changes

ben9583 commented:
This is looking pretty much all good, just a couple notes

- The post IDs should be cached too (in addition to the text which you've done), so that way we don't have to call the GitHub API to scan all of the announcements so often
- There should be `get_last_n_announcements` that just returns the IDs and metadata (using caching), we don't need to get all the last `n` announcement full text contents

Let me know if you have questions of what I mean by this cause I may have worded this a bit weird. I can explain more on specifics

vicshi06 commented:
I have changed the structure a little. The script now checks for updated cache whenever `get_last_n_announcements` and `get_last_n_announcements_text` are called, so the cache checking behavior is more consistent. 

I have also updated the `README` so people would know how to test individual test file. 

Let me know if you have any thoughts! The script should be good to go. I haven't added the tests for the cache since I have not yet thought of a good way to test it. It might make more sense to actually call it from the front end to make sure its behavior is good. But it is not super complex so we don't necessarily need super complex tests.
